File: lc/py/049.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        def rob1(start, stop):
            prev = 0
            adj = 0
            for i in range(start, stop):
                max_val = max(adj, nums[i]+prev)
                prev = adj
                adj = max_val
            return adj

        if len(nums) == 1:
            return nums[0]

        return max(rob1(0, len(nums)-1), rob1(1, len(nums)))

File: lc/c/049.c
int rob1(int *nums, int start, int stop) {
    int prev = 0, adj = 0;
    for (int i = start; i < stop; ++i) {
        int max_val = prev+nums[i] > adj ? prev+nums[i] : adj;
        prev = adj;
        adj = max_val;
    }
    return adj;
}

int rob(int* nums, int numsSize){
    if (numsSize == 1)
        return nums[0];

    int max1 = rob1(nums, 0, numsSize-1);
    int max2 = rob1(nums, 1, numsSize);
    return max1 > max2 ? max1 : max2;
}
